cargo-dist/templates/installer/installer.sh.j2

```diff
@@ -332,48 +368,41 @@ install() {
-    # In this case we want to be late-bound, as $HOME is reliable/nice.
-    if [ -n "${HOME:-}" ]; then
-        _install_dir="$HOME/{{ install_path.subdir }}"
-        _env_script_path="$HOME/{{ install_path.subdir }}/env"
-        _install_dir_expr='$HOME/{{ install_path.subdir }}'
-        _env_script_path_expr='$HOME/{{ install_path.subdir }}/env'
-
-        if [ -n "${CARGO_DIST_FORCE_INSTALL_DIR:-}" ]; then
-            _install_dir="$CARGO_DIST_FORCE_INSTALL_DIR"
-            _env_script_path="$CARGO_DIST_FORCE_INSTALL_DIR/env"
-            # For these two, if the new value contains $HOME, make
-            # sure to `sed` the $HOME expression back into the expr
-            _install_dir_expr="$(echo "$CARGO_DIST_FORCE_INSTALL_DIR" | sed "s,$HOME,\$HOME,")"
-            _env_script_path_expr="$(echo "$CARGO_DIST_FORCE_INSTALL_DIR/env" | sed "s,$HOME,\$HOME,")"
-        fi
+    _install_dir="$HOME/$INSTALL_PATH_SUBDIR"
+    _env_script_path="$HOME/$INSTALL_PATH_SUBDIR/env"
+    _install_dir_expr="\$HOME/$INSTALL_PATH_SUBDIR"
+    _env_script_path_expr="\$HOME/$INSTALL_PATH_SUBDIR/env"
+
+    if [ -n "${CARGO_DIST_FORCE_INSTALL_DIR:-}" ]; then
+        _install_dir="$CARGO_DIST_FORCE_INSTALL_DIR"
+        _env_script_path="$CARGO_DIST_FORCE_INSTALL_DIR/env"
+        # For these two, if the new value contains $HOME, make
+        # sure to `sed` the $HOME expression back into the expr
+        _install_dir_expr="$(echo "$CARGO_DIST_FORCE_INSTALL_DIR" | sed "s,$HOME,\$HOME,")"
+        _env_script_path_expr="$(echo "$CARGO_DIST_FORCE_INSTALL_DIR/env" | sed "s,$HOME,\$HOME,")"
+    fi
+elif [ "$INSTALL_PATH_STRATEGY" = "EnvSubdir" ]; then
+    # Install to this subdir of the user's $INSTALL_PATH_ENV_KEY dir.
+    if [ -n "$INSTALL_PATH_SUBDIR" ]; then
+        _install_dir="$INSTALL_PATH_ENV_KEY/$INSTALL_PATH_SUBDIR"
+        _env_script_path="$INSTALL_PATH_ENV_KEY/$INSTALL_PATH_SUBDIR/env"
     else
-        err "could not find your HOME dir to install binaries to"
+        _install_dir="$INSTALL_PATH_ENV_KEY"
+        _env_script_path="$INSTALL_PATH_ENV_KEY/env"
     fi
-{% elif install_path.kind == "EnvSubdir" %}
-    # Install to this subdir of the user's {{ install_path.env_key }} dir.
-    # In this case we want to be early-bound, as the env-var can't be trusted longterm.
-    if [ -n "{{ "${" }}{{ install_path.env_key }}:-}" ]; then
-        _install_dir="${{ install_path.env_key }}{% if install_path.subdir | length %}/{% endif %}{{ install_path.subdir }}"
-        _env_script_path="${{ install_path.env_key }}{% if install_path.subdir | length %}/{% endif %}{{ install_path.subdir }}/env"
+    _install_dir_expr="$_install_dir"
+    _env_script_path_expr="$_env_script_path"
+
+    # Override if necessary
+    if [ -n "${CARGO_DIST_FORCE_INSTALL_DIR:-}" ]; then
+        _install_dir="$CARGO_DIST_FORCE_INSTALL_DIR"
+        _env_script_path="$CARGO_DIST_FORCE_INSTALL_DIR/env"
         _install_dir_expr="$_install_dir"
```

Comment:
I think this is the fatal issue with this implementation: the install strategy that resolves *today* is completely unrelated to the install strategy that resolved when the app was first installed.

And I don't mean because a developer can change their cargo-dist config. I mean because and end-user's env-vars have to be regarded as ephemeral. We *specifically* allow $HOME to remain dynamic, but $MYCOMPANY is fully "substituted" during installation, on the assumption that it was temporarily set and won't be around for long.

As a result, the updater needs to trust the install script to behave correctly if it (re)runs an install script with different env-vars set. So all handling of CARGO_DIST_FORCE_INSTALL_DIR has to be unconditional, and not nested within any particular INSTALL_PATH_STRATEGY.

This is why I was pushing for the normalization stuff before doing this change -- it would make that trivially true.

This is also why I suggested a partial compromise: only allow CARGO_HOME in install-strategy if it's the *only* option. That way there's two cases for CARGO_DIST_FORCE_INSTALL_DIR:

* either it's *just* CARGO_HOME and we have to append /bin
* or it's an arbitrary pile of $ENVVAR and $HOME checks which use the same flat structure

in either case we get to emit uniform CARGO_DIST_FORCE_INSTALL_DIR logic

(Note: my proposal *will* result in us substituting $HOME more aggressively on re-installs if the install was done with an env-var, which in the current implementation will add a duplicate entry to rcfiles. This is not fatal just ugly. We could fix this by more checking for more variations when checking to see if we should edit rcfiles.)